pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+		}
+
+	}
+	glog.V(2).Infof("Gluster server and Backup servers [%+v,%+v]", glusterServer, bkpservers)
+
+	return glusterServer, bkpservers, err
+}
+
+// DeleteVolume deletes the given volume. The function is idempotent.
+func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+        // TODO; return success if volume not found
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Volume delete request is nil")
+	}
+
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "volume ID is nil")
+	}
+	glog.V(2).Infof("Deleting volume with ID: %v", req.VolumeId)
+
+	err := cs.client.VolumeStop(req.VolumeId)
+
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to stop volume %s", err.Error())
+	}
+
+	err = cs.client.VolumeDelete(req.VolumeId)
+	if err != nil {
+		glog.Errorf("Volume delete failed :%v", err)
+		return nil, status.Errorf(codes.Internal, "error deleting volume: %s", err.Error())
+	}
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+// ControllerPublishVolume attaches the given volume to the node
+func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+//ControllerUnpublishVolume deattaches the given volume from the node
+func (cs *ControllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+// ValidateVolumeCapabilities checks whether the volume capabilities requested
+// are supported.
+func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "volume capabilities request is nil")
+	}
+
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ValidateVolumeCapabilities() - Volume ID is nil")
+	}
+
+	if req.VolumeCapabilities == nil {
+		return nil, status.Error(codes.InvalidArgument, "ValidateVolumeCapabilities is nil")
+	}
+
+	var vcaps []*csi.VolumeCapability_AccessMode
+	for _, mode := range []csi.VolumeCapability_AccessMode_Mode{
+		csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
+	} {
+		vcaps = append(vcaps, &csi.VolumeCapability_AccessMode{Mode: mode})
+	}
+
+	hasSupport := func(mode csi.VolumeCapability_AccessMode_Mode) bool {
+		for _, m := range vcaps {
+			if mode == m.Mode {
+				return true
+			}
+		}
+		return false
+	}
+
+	resp := &csi.ValidateVolumeCapabilitiesResponse{
+		Supported: false,
+	}
+
+	for _, cap := range req.VolumeCapabilities {
+		if hasSupport(cap.AccessMode.Mode) {
+			resp.Supported = true
+		} else {
+			resp.Supported = false
+		}
+	}
```

Comment:
This is buggy. We would be returning `Supported = true` when the last requested capability is supported even if earlier capabilities are not. You should be breaking out when any requested capability is unsupported.